sql/analyzer/apply_join.go

```diff
@@ -245,3 +246,153 @@ func normalizeSelectSingleRel(ctx *sql.Context, a *Analyzer, n sql.Node, scope *
+//
+// select * from xy where exists (select * from uv where x = 1)
+// =>
+// select * from xy where x = 1 and exists (select * from uv)
+func hoistOutOfScopeFilters(ctx *sql.Context, a *Analyzer, n sql.Node, scope *Scope, sel RuleSelector) (sql.Node, transform.TreeIdentity, error) {
+	switch n.(type) {
+	case *plan.TriggerBeginEndBlock:
+		return n, transform.SameTree, nil
+	}
+	ret, same, _, err := recurseSubqueryForOuterFilters(n, a, scope)
+	return ret, same, err
+}
+
+// recurseSubqueryForOuterFilters recursively hoists filters that belong
+// to an outer scope (maybe higher than the parent). We do a DFS for hoisting
+// subquery filters. We do a BFS to extract hoistable filters from subquery
+// expressions before checking the normalized subquery and its hoisted
+// filters for further hoisting.
+func recurseSubqueryForOuterFilters(n sql.Node, a *Analyzer, scope *Scope) (sql.Node, transform.TreeIdentity, []sql.Expression, error) {
+	var hoistFilters []sql.Expression
+	lowestAllowedIdx := len(scope.Schema())
+	var inScope TableAliases
+	ret, same, err := transform.Node(n, func(n sql.Node) (sql.Node, transform.TreeIdentity, error) {
+		sq, _ := n.(*plan.SubqueryAlias)
+		if sq != nil {
+			subScope := scope.newScopeFromSubqueryAlias(sq)
+			newQ, same, hoisted, err := recurseSubqueryForOuterFilters(sq.Child, a, subScope)
+			if err != nil {
+				return n, transform.SameTree, err
+			}
+			if same {
+				return n, transform.SameTree, nil
+			}
+			if len(hoisted) > 0 {
+				hoistFilters = append(hoistFilters, hoisted...)
+			}
+			ret := *sq
+			ret.Child = newQ
+			return &ret, transform.NewTree, nil
+		}
+		f, _ := n.(*plan.Filter)
+		if f == nil {
+			return n, transform.SameTree, nil
+		}
+
+		var keepFilters []sql.Expression
+		allSame := transform.SameTree
+		queue := splitConjunction(f.Expression)
+		for len(queue) > 0 {
+			e := queue[0]
+			queue = queue[1:]
+
+			var not bool
+			if n, ok := e.(*expression.Not); ok {
+				not = true
+				e = n.Child
+			}
+
+			// (1) normalize subquery expressions
+			// (1a) recurse downwards
+			// (1b) add hoisted to queue
+			// (1c) standardize subquery expression for hoisting
+			var sq *plan.Subquery
+			switch e := e.(type) {
+			case *plan.InSubquery:
+				sq, _ = e.Right.(*plan.Subquery)
+			case *plan.ExistsSubquery:
+				sq = e.Query
+			default:
+			}
+			if sq != nil {
+				children := e.Children()
+				subScope := scope.newScopeFromSubqueryExpression(n)
+				newQ, same, hoisted, err := recurseSubqueryForOuterFilters(sq.Query, a, subScope)
+				if err != nil {
+					return n, transform.SameTree, err
+				}
+				allSame = allSame && same
+				if len(hoisted) > 0 {
+					newScopeFilters, _, err := FixFieldIndexesOnExpressions(scope, a, n.Schema(), hoisted...)
+					if err != nil {
+						return n, transform.SameTree, err
+					}
+					queue = append(queue, newScopeFilters...)
+				}
+				newSq := sq.WithQuery(newQ)
+				children[len(children)-1] = newSq
```

Comment:
`children[len(children)-1] = newSq` works because `e` is either:
- `ExistsSubquery`, and in that case we're updating `e.Query`, which is the only/last child
- `InSubquery`, and in that case we're updating `e.Right`, which is the last child

Two comments:
1. How do we enforce that `InSubquery.Right` is of type `plan.Subquery`? Is it possible that `e.Left` will contain `plan.Subquery`?
2. `children[len(children)-1]` is a bit hard to reason about and makes the above assumption. I wonder if it would make sense to loop over `e.Children()` to find the `plan.Subquery`. This way we will always have the correct index of the child and won't have to make assumptions.

Reply:
`InSubquery.Right` and `ExistsSubquery.Child` have to be subqueries. I have tried changing the nodes to enforce this type restriction, but it ends up refactoring hundreds-~thousand lines of code, so for now we just leave the type conversion.